ocd_backend/pipeline.py

```diff
@@ -52,46 +53,113 @@ def setup_pipeline(source_definition):
-                # from an extractor have finished
+    # we can have multiple pipelines. but for compatibility and readability
+    # use the source definition if no specific pipelines have been defined
+    pipelines = source_definition.get('pipelines', None) or [source_definition]
+
+    pipeline_definitions = {}
+    pipeline_extractors = {}
+    pipeline_transformers = {}
+    pipeline_enrichers = {}
+    pipeline_loaders = {}
+
+    for pipeline in pipelines:
+        if 'id' not in pipeline:
+            raise ConfigurationError("Each pipeline must have an id field.")
+
+        # adjusted source definitionsv per pipeline. This way you can for
+        # example change the index on a pipeline basis
+        pipeline_definitions[pipeline['id']] = deepcopy(source_definition)
+        pipeline_definitions[pipeline['id']].update(pipeline)
+
+        # initialize the ETL classes, per pipeline
+        pipeline_extractors[pipeline['id']] = [
+            load_object(cls) for cls in
+            pipeline_definitions[pipeline['id']].get('extractors', None) or [
+                pipeline_definitions[pipeline['id']]['extractor']
+            ]
+        ]
+
+        pipeline_transformers[pipeline['id']] = load_object(
+            pipeline['transformer'])()
+
+        pipeline_enrichers[pipeline['id']] = [
+            (load_object(enricher[0])(), enricher[1]) for enricher in
+            pipeline_definitions[pipeline['id']].get('enrichers', [])]
+
+        pipeline_loaders[pipeline['id']] = [
+            load_object(cls)() for cls in
+            pipeline_definitions[pipeline['id']].get('loaders', None) or [
+                pipeline_definitions[pipeline['id']]['loader']
+            ]
+        ]
+
+    for pipeline in pipelines:
+        try:
+            # The first extractor should be a generator instead of a task
+            for item in pipeline_extractors[pipeline['id']][0](
+                    source_definition=pipeline_definitions[pipeline['id']]).run():
+
+                step_chain = chain()
 
                 params['chain_id'] = uuid4().hex
-                celery_app.backend.add_value_to_set(set_name=run_identifier_chains,
-                                                    value=params['chain_id'])
-
-                step_class = load_object(step)()
-
-                step_chain |= step_class.s(
+                celery_app.backend.add_value_to_set(
+                    set_name=run_identifier_chains,
+                    value=params['chain_id'])
+
+                # Remaining extractors
+                for extractor in pipeline_extractors[pipeline['id']][1:]:
+                    step_chain |= extractor().s(
+                        *item,
+                        source_definition=pipeline_definitions[pipeline['id']],
+                        **params
+                    )
+                    # Prevent old item being passed down to next steps
+                    item = []
```

Comment:
This fixes the problem of giving an old item to a following extractor or transformer. I am not sure if there is a better way to do this.